plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java

```diff
@@ -16,96 +16,108 @@
-   * may be presented, e.g., if no user is logged in or login is required due to an expired
-   * credential. This method returns {@code null} if a user cancels the login process.
-   * For this reason, if {@code null} is returned, the caller should cancel the current
-   * operation and display a general message that login is required but was cancelled or failed.
-   *
-   * Must be called from a UI context.
-   *
-   * @param shellProvider provides a shell for the login screen if login is necessary
-   * @throws IOException can be thrown by the underlying Login API request (e.g., network
-   *     error from the transport layer while sending/receiving a HTTP request/response.)
-   */
-  public Credential getActiveCredential(IShellProvider shellProvider) throws IOException {
-    Credential credential = getCachedActiveCredential();
-
-    if (credential == null) {
-      credential = logIn(shellProvider);
-
-      IEclipseContext eclipseContext = PlatformUI.getWorkbench().getService(IEclipseContext.class);
-      eclipseContext.set(STASH_OAUTH_CRED_KEY, credential);
-    }
-    return credential;
+  private static final SortedSet<String> OAUTH_SCOPES = Collections.unmodifiableSortedSet(
+      new TreeSet<>(Arrays.asList(
+          "email", //$NON-NLS-1$
+          "https://www.googleapis.com/auth/cloud-platform" //$NON-NLS-1$
+      )));
+
+  private GoogleLoginState loginState;
+  private AtomicBoolean loginInProgress;
+
+  @VisibleForTesting
+  GoogleLoginService(
+      OAuthDataStore dataStore, UiFacade uiFacade, LoggerFacade loggerFacade) {
+    loginState = new GoogleLoginState(
+        Constants.getOAuthClientId(), Constants.getOAuthClientSecret(), OAUTH_SCOPES,
+        dataStore, uiFacade, loggerFacade);
+    loginInProgress = new AtomicBoolean(false);
   }
 
-  /**
-   * Returns the credential of an active user (among multiple logged-in users). Unlike {@link
-   * #getActiveCredential}, this version does not involve login process or make API calls.
-   * Returns {@code null} if no credential has been cached.
-   *
-   * Safe to call from non-UI contexts.
-   */
-  public Credential getCachedActiveCredential() {
-    IEclipseContext eclipseContext = PlatformUI.getWorkbench().getService(IEclipseContext.class);
-    return (Credential) eclipseContext.get(STASH_OAUTH_CRED_KEY);
+  public GoogleLoginService() {
+    this(new TransientOAuthDataStore(PlatformUI.getWorkbench().getService(IEclipseContext.class)),
+        new LoginServiceUi(), new LoginServiceLogger());
   }
 
-  private Credential logIn(IShellProvider shellProvider) throws IOException {
-    GoogleLoginBrowser loginBrowser = new GoogleLoginBrowser(
-        shellProvider.getShell(), Constants.getOAuthClientId(), OAUTH_SCOPES);
-    if (loginBrowser.open() != GoogleLoginBrowser.OK) {
-      return null;
+  @Override
+  public Credential getActiveCredential() {
+    if (!loginInProgress.compareAndSet(false, true)) {
+      LoginServiceUi.showErrorDialogHelper(
```

Comment:
Ah, the problem was that, because of [this](https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/486/files/535f3b5b0506cf3fab62fd9015202df008d4e085#r73581984), `title` and `message` are completely ignored in `UiFacade#showErrorDialog(title,message)`. The method is not usable.